Add `ShapeCast2D` node

Xrayez opened the pull request:
Implements the 2D part of godotengine/godot-proposals#710.

![image](https://user-images.githubusercontent.com/17108460/140937094-06807d4f-32ba-4750-a496-3d0eac7f9029.png)

Ported from [Goost](https://goostengine.github.io/) to Godot 4.0, implementation mostly equivalent to https://github.com/goostengine/goost/blob/46379f65c5bc9324dc0be1cb9c13f34c22b681b2/scene/physics/2d/shape_cast_2d.cpp, except collisions are processed regardless whether shape cast is already colliding or not.

![image](https://user-images.githubusercontent.com/17108460/140936766-70946eb1-3a4c-432b-87fc-7f9020ca4f81.png)

Drawing could be improved, but perhaps that something @pouleyKetchoupp could build upon, as well as implementing the 3D counterpart based on this.

**Test project**:
[shape-cast-2d.zip](https://github.com/godotengine/godot/files/7521544/shape-cast-2d.zip)

*(left mouse button click to set shape cast global position, right click to set target position)*

Commit:
Add `ShapeCast2D` node

pouleyKetchoupp commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,466 @@
+	return collision_mask;
+}
+
+void ShapeCast2D::set_collision_mask_bit(int p_bit, bool p_value) {
```
This method has been renamed in RayCast2D to `set_collision_mask_value()` and it takes a 1-based value now. Please copy the one from RayCast2D to make it consistent.

pouleyKetchoupp commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,466 @@
+	set_collision_mask(mask);
+}
+
+bool ShapeCast2D::get_collision_mask_bit(int p_bit) const {
```
Same with `get_collision_mask_value()`.

pouleyKetchoupp commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,466 @@
+
+		case NOTIFICATION_DRAW: {
+#ifdef TOOLS_ENABLED
+			if (!Engine::get_singleton()->is_editor_hint() && !get_tree()->is_debugging_collisions_hint()) {
```
Missing error `ERR_FAIL_COND(!is_inside_tree());` recently added in RayCast2D to make it safe.

pouleyKetchoupp commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,466 @@
+		col["collider_id"] = sri.collider_id;
+		col["shape"] = sri.shape;
+		col["linear_velocity"] = sri.linear_velocity;
+		// col["metadata"] = sri.metadata;
```
You can remove this line completely, metadata is not supported anymore.

Xrayez replied:
Is there a replacement/alternative to metadata in Godot 4.x? It's needed for determining which `TileMap` quadrants get affected by collision detection in Godot 3.x (see use case at /godotengine/godot-proposals#2543).

pouleyKetchoupp replied:
Yes, it should be even a bit more clear since it's documented now.

In 4.0 each tile has a separate body, and you can access the corresponding tile coordinates based on the body id directly from the tilemap:
https://github.com/godotengine/godot/blob/284b50cf30a939dfee91876e06895aa50940c37c/doc/classes/TileMap.xml#L102-L108

MNchino replied:
> Yes, it should be even a bit more clear since it's documented now.
> 
> In 4.0 each tile has a separate body, and you can access the corresponding tile coordinates based on the body id directly from the tilemap:
> 
> https://github.com/godotengine/godot/blob/284b50cf30a939dfee91876e06895aa50940c37c/doc/classes/TileMap.xml#L102-L108

Is there a way to get the tile's layer as well?
In order to retrieve custom data from the tile (from the custom data layer), we need to know the tile atlas coords ( aka because We need to call get_tile_data func on a TilesetAtlasSource, and then call get_custom_data on the TileData).

This is needed for any system that wants to identify the tile in a collision while also using layers.  Otherwise, all tiles will need to live in one layer or a known layer of the collision source

pouleyKetchoupp commented on scene/2d/shape_cast_2d.h:
```diff
@@ -0,0 +1,126 @@
+
+	TypedArray<String> get_configuration_warnings() const override;
+
+	ShapeCast2D(){};
```
Empty constructor not needed.

pouleyKetchoupp reviewed (changes requested):
There are a few details to fix before merging, but otherwise it looks great. Thanks!

A few more things can be improved but they don't have to be part of this PR. I'm leaving a list here for the record:
- Improve debug draw by making a convex hull around the whole path
- Add a shape transform property to allow rotation/scale without changing the cast
- Add an equivalent node for 3D (still needs some tests to decide how to draw it)

Xrayez commented:
> There are a few details to fix before merging, but otherwise it looks great. Thanks!

Done!

> Improve debug draw by making a convex hull around the whole path

I have no idea what kind of drawing capabilities there are in 4.x, but perhaps there's a way to simply use an appropriate blending mode. Otherwise convex hull will require to refactor `Shape2D.draw()` into `Shape2D.get_points()` + `draw()`.

> Add a shape transform property to allow rotation/scale without changing the cast

I've tried to implement something like this myself but couldn't find a way to do this on the front-end side, I think shape transform should be done during `cast_motion()/rest_info()`, especially if there's a possibility to interpolate *rotation* between origin and target positions to simulate *angular* motion, not just linear, but this is probably too much for Godot (?)

In either case, yeah it would be nice to merge this sooner so that we don't lose motivation to work on all this stuff!

fire commented:
There is a deluanlay tetrahedralize for 3d points can the 2d version be used?

Xrayez commented:
There's `Geometry2D.convex_hull()` or similar, so it should be possible already. But we'd need to either expose `Shape2D` points for each shape, or hardcode point generation for each shape locally in `ShapeCast2D`. I'd prefer to expose `Shape2D.get_points()` in order to get geometrical representation of the shape via polygon vertices, because `Shape2D.draw()` already generates points internally for them to be drawn.

We could keep `Shape.get_points()` for internal use, but there are several proposals at Godot that could benefit from this, such as godotengine/godot-proposals#3495 or godotengine/godot-proposals#1925.

But this is something which touches core parts of the engine, so it's kind of out of scope for this PR, unless you think that implementing `Shape2D.get_points()` for each shape makes sense.

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+				[b]Note:[/b] [code]enabled == true[/code] is not required for this to work.
+			</description>
+		</method>
+		<method name="get_closest_collider" qualifiers="const">
```
The `get_closest_` functions are not going to return correct results until some changes are made in the physics server API, so better remove them for now (except for safe/unsafe distances).

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+				[b]Note:[/b] this point is in the [b]global[/b] coordinate system.
+			</description>
+		</method>
+		<method name="get_closest_collision_safe_distance" qualifiers="const">
```
The physics server actually returns [0..1] fractions, not distances, so either these functions should be renamed or the distance needs to be calculated from the motion length.

Xrayez replied:
`get_closest_collision_safe_fraction` or `get_closest_collision_safe_motion_fraction()` then? Kind of verbose though.

pouleyKetchoupp replied:
It should be either `get_closest_collision_safe_fraction()` and a clearer explanation that it gives fractions between 0 and 1, or `get_closest_collision_safe_distance()` and in this case the method needs to be updated to return `motion_length * fraction`.

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+			<return type="Object" />
+			<argument index="0" name="index" type="int" />
+			<description>
+				Returns the first object that the shape intersects, or [code]null[/code] if no object is intersecting the shape (i.e. [method is_colliding] returns [code]false[/code]).
```
Rewording to match other functions:
```suggestion
				Returns the [Object] of one of the multiple collisions at [code]index[/code] that the shape intersects, or [code]null[/code] if no object is intersecting the shape (i.e. [method is_colliding] returns [code]false[/code]).
```

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+	</brief_description>
+	<description>
+		Shape casting allows to detect collision objects along the cast area (for instance, beam weapons).
+		Immediate collision overlaps can be easily done with the [member target_position] set to [code]Vector2(0, 0)[/code] and by calling [method force_shapecast_update] within the same [b]physics_frame[/b]. This also helps to overcome some limitations of [Area2D] when used as a continuos detection area, often requiring waiting a couple of frames before collision information is available to [Area2D] n ...
```
Typo:
```suggestion
		Immediate collision overlaps can be easily done with the [member target_position] set to [code]Vector2(0, 0)[/code] and by calling [method force_shapecast_update] within the same [b]physics_frame[/b]. This also helps to overcome some limitations of [Area2D] when used as a continuous detection area, often requiring waiting a couple of frames before collision information is available to [Area2D] nodes, and when using the signals creates unnecessary complexity.
```

pouleyKetchoupp commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,467 @@
+	params.collide_with_areas = collide_with_areas;
+
+	if (target_position != Vector2()) {
+		dss->cast_motion(params, collision_safe_distance, collision_unsafe_distance);
```
Safe and unsafe distances could be initialized to 0 when the motion is zero to keep things consistent.

pouleyKetchoupp reviewed (changes requested):
I've left more comments after testing the ShapeCast2D node.

It generally works well, but there are some inconsistencies between what the documentation states and the results you get, so I've tried to suggest minimal changes to make it consistent until this node can be improved along with the physics API.

Xrayez commented on scene/2d/shape_cast_2d.cpp:
```diff
@@ -0,0 +1,467 @@
+	bool intersected = true;
+	while (intersected && result.size() < max_results) {
+		PhysicsDirectSpaceState2D::ShapeRestInfo info;
+		intersected = dss->rest_info(params, &info);
```
I've found `rest_info()` to be the best compromise in 3.x, because:

1. You can detect multiple bodies along the cast given detected bodies get subsequently excluded from intersection:

![image](https://user-images.githubusercontent.com/17108460/141168946-0712b295-9813-4fc5-a619-35cc2b0358a8.png)

2. It does return useful information such as normal.

pouleyKetchoupp replied:
I agree the position and normal would be good to have, but given the current API it just returns collisions in a random order and the positions are not correct so it's not really usable in common cases.

What you're trying to do makes sense. I would just rather wait for making the proper changes in the physics API before exposing the extra information to avoid confusing users, since it's not working as intended for now.

Xrayez replied:
What if we keep `rest_info()` when `motion == Vector2()`, and use `intersect_shape()` when `motion != Vector2()`? Would that be a good compromise?

pouleyKetchoupp replied:
Yes, that could be a good compromise as long as it's clear and well documented. It does sound like a better solution than using a separate node.

Xrayez replied:
It was all wrong from the start due to porting error, see https://github.com/godotengine/godot/pull/54803#issuecomment-966407480.

Xrayez commented:
> Unfortunately there's currently no good API that gives you all the information you need for all collisions.

By the way, given above suggestions, I'm fine for this PR to be postponed and be used as a reference point until physics backend is improved to match the proposed API, to be honest. I wouldn't want to remove those `closest_*` methods and other features just for them to be implemented again, nor anybody else would want to do double work, I presume.

pouleyKetchoupp commented:
@Xrayez Ok, that makes sense! What I can propose is to make the adjustments to the physics server API in the next two weeks or so, and then you can update this PR to have the basic functionality work as intended and we can merge it.

After that, other contributors will be able to help with adding functionalities and improving the rendering.

And for now, this PR can be in draft mode until the API is ready.

Xrayez commented:
@pouleyKetchoupp @djrain so I thought I was losing my mind while porting the node to Godot 4.0 in godotengine/godot#54803, but I've checked the implementation in Goost I've made a year ago for Godot 3.x and figured that:
* `rest_info()` *does not* take into account `motion` (hardcoded to `Vector2()`).
* Therefore, I manually *translate* the original global transform to the point of unsafe motion fraction like this in 3.x right after `cast_motion()`:

```c++
global_transform.set_origin(global_transform.get_origin() + to * collision_unsafe_fraction);
```

Only then I use `rest_info()`, but only at the point of impact (or when the shape is already stuck), and I'm not using `intersect_shape()` because unlike `rest_info()`, it doesn't return normal/contact info etc.

I have now properly ported implementation from 3.x to 4.x, it should work alright now (hopefully), sorry for confusion as I got confused by current API in Godot 4.0 while porting this. I have updated the test project according to suggested renames as well.

pouleyKetchoupp commented:
Ok, the contact positions seem to work, although now it always only returns one collision result even if the shape cast goes through several bodies/areas. In order to get exactly the intended result, maybe you would need to use `cast_motion()` again for each iteration? Otherwise it just tests for rest info several times at the first hit position.

Here's the project I'm using for tests:
[shape-cast-4.0.zip](https://github.com/godotengine/godot/files/7523849/shape-cast-4.0.zip)

Xrayez commented:
The "intended" result I've shown at https://github.com/godotengine/godot/pull/54803#discussion_r746859107 wasn't actually intended, sorry!

The intended result was to hit a single object at the time of impact, but potentially return multiple collisions at point of impact in case there are multiple bodies there. This is consistent with `RayCast2D`, which doesn't support detecting multiple collisions *along the cast motion*, see proposal #25695.

The "closest" object in this case would be the one that first intersects with `rest_info()` in static case (if multiple bodies are also detected at that position, not just with the body).

If you'll say that `rest_info()` wouldn't return closest object at *static* collision in some cases, then I think `get_closest_*` methods could be renamed to `get_first_impact_collision()` or something like that.

Due to above, and if #25695 is implemented, then `RayCast2D` would also need to introduce `get_closest_*` methods in order to differentiate those bodies which are detected first, and bodies that are detected subsequently, see also reduz proposed solution: https://github.com/godotengine/godot/issues/14608#issuecomment-354024062 (which I already do in the case of `ShapeCast2D`, but only in order to detect multiple bodies at time of impact, and not multiple bodies along the cast motion, which I feel like asks for another pull requests, since `RayCast2D` doesn't support this either at the moment).

That said, it's always possible to move `ShapeCast2D/RayCast2D` at the point of impact after the collision via script, and repeat the process.

Also, in your test project, `ShapeCast2D` was configured to collide with `Area2D`, while `RayCast2D` wasn't, which lead to different result.

pouleyKetchoupp commented:
> The intended result was to hit a single object at the time of impact, but potentially return multiple collisions at point of impact in case there are multiple bodies there. This is consistent with `RayCast2D`, which doesn't support detecting multiple collisions _along the cast motion_, see proposal #25695.

Alright, makes sense to make it consistent with `RayCast2D`, and add extra features to both later. It all seems to work well, the only thing I noticed is if you increase the margin, rest info can fail. But I think I've found where the bug is in the physics server, I'll fix it in a separate PR (rest info can fail to return anything if margin is too high and there's no motion).

> The "closest" object in this case would be the one that first intersects with rest_info() in static case (if multiple bodies are also detected at that position).
>
> If you'll say that rest_info() wouldn't return closest object at static collision in some cases, then I think get_closest_* methods could be renamed to get_first_impact_collision() or something like that.

Honestly, I would rather remove the `get_closest_*` methods because whatever their name is it seems they might just add some confusion. On the other hand, the `get_*` methods that use a collision index could be clarified with some wording like `one of the collisions at impact`.

I'm going to add more review comments for some details too, I think it's almost ready but I'd like to make sure the documentation is clear.

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+			<return type="int" />
+			<argument index="0" name="index" type="int" />
+			<description>
+				Returns the [Object] of one of the multiple collisions at [code]index[/code] that the shape intersects, or [code]0[/code] if no object is intersecting the shape (i.e. [method is_colliding] returns [code]false[/code]).
```
It looks like you might have copied the text meant for `get_collider()` in the wrong section by mistake.

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+		<method name="get_collision_count" qualifiers="const">
+			<return type="int" />
+			<description>
+				The number of collisions detected. Use this to iterate over multiple collisions as provided by [method get_collider], [method get_collider_shape], [method get_collision_point], and [method get_collision_normal] methods.
```
In this method too it could be a bit clearer:
```suggestion
				The number of collisions detected at the point of impact. Use this to iterate over multiple collisions as provided by [method get_collider], [method get_collider_shape], [method get_collision_point], and [method get_collision_normal] methods.
```

Xrayez commented:
> Honestly, I would rather remove the `get_closest_*` methods because whatever their name is it seems they might just add some confusion. On the other hand, the `get_*` methods that use a collision index could be clarified with some wording like `one of the collisions at impact`.

That's good enough for me as long as multiple collisions are able to be detected with contact info.

pouleyKetchoupp commented on doc/classes/ShapeCast2D.xml:
```diff
@@ -0,0 +1,182 @@
+			If [code]true[/code], the parent node will be excluded from collision detection.
+		</member>
+		<member name="margin" type="float" setter="set_margin" getter="get_margin" default="0.0">
+			The collision margin for the shape. Large margin speeds up the performance at the cost of precision.
```
I'm not sure the margin actually speeds up performance in this case. In the case of character controllers it's mostly used to allow collision detection to be a bit more flexible so floor and wall detection can be more stable, but I'm not sure if there are other use cases for shape cast margin.

So maybe it could be something like:
```suggestion
			The collision margin for the shape. A larger margin helps detecting collisions more consistently, at the cost of precision.
```

pouleyKetchoupp reviewed (changes requested):
Here are a few last things that can be made clearer in the documentation. The code itself looks fine to me!

djrain commented:
Glad I read this before going any further with my gdscript shapecasting system 😅

I have something working well enough for now, so I'm wondering @pouleyKetchoupp what changes you currently have in mind for the physics API and when/where they might actually be merged (if you have any idea)?

Commit:
Remove `get_closest_*` methods from `ShapeCast2D`

The physics API cannot provide needed functionality to ensure the correct behavior, which might lead to confusion (see `rest_info()`).

However `get_closest_collision_safe/unsafe_fraction()` methods are not removed, because they return correct result from `cast_motion()`.

Xrayez commented:
> > Honestly, I would rather remove the `get_closest_*` methods because whatever their name is it seems they might just add some confusion. On the other hand, the `get_*` methods that use a collision index could be clarified with some wording like `one of the collisions at impact`.
> 
> That's good enough for me as long as multiple collisions are able to be detected at time of impact with contact info such as normals (that will be mostly useful for cases when shape cast starts in a colliding/stuck state at origin against other collision objects).

I have pushed a separate commit to document this, because that's what differs from Goost implementation now, and that's something to keep in mind when new/updated API is planned for physics regarding dynamic collisions in general.

pouleyKetchoupp commented:
> I have something working well enough for now, so I'm wondering @pouleyKetchoupp what fixes/changes you currently have in mind for the physics API and when/where they might actually be merged (if you have any idea)?

I was thinking of making the different shape functions more consistent with each other, and allow all of them to return proper collision information (for example `cast_motion` currently returns just the fraction and maybe some others can be improved too).

Given the state of this PR, it's not really needed anymore, but it might help with optimizing the queries, especially if we want to support getting different results along the motion later.

Since it might include some compatibility breaking changes, I'll try to make a PR in the next few weeks, but I can't really give you a precise date.

pouleyKetchoupp reviewed (approved):
Looks good to go!
